clang/include/clang/CIR/Dialect/IR/CIROps.td

```diff
@@ -2446,4 +2446,41 @@ def AssumeOp : CIR_Op<"assume"> {
   }];
 }
 
+//===----------------------------------------------------------------------===//
+// Branch Probability Operations
+//===----------------------------------------------------------------------===//
+
+def ExpectOp : CIR_Op<"expect",
+  [Pure, AllTypesMatch<["result", "val", "expected"]>]> {
+  let summary = "Tell the optimizer that two values are likely to be equal.";
+  let description = [{
+    The `cir.expect` operation may take 2 or 3 arguments.
+
+    When the argument `prob` is missing, this operation effectively models the
+    `__builtin_expect` builtin function. It tells the optimizer that `val` and
+    `expected` are likely to be equal.
+
+    When the argumen `prob` is present, this operation effectively models the
+    `__builtin_expect_with_probability` builtin function. It tells the
+    optimizer that `val` and `expected` are equal to each other with a certain
+    probability.
+
+    `val` and `expected` must be integers and their types must match.
```

Comment:
It looks like in practice these will only be i64, since that's what the __builtin_expect functions take as arguments, but I like the idea of being able to reduce the size to eliminate casts during optimization, and it seems like it would be useful to be able to have bool arguments as well. I'm not saying this needs to change now. I'm just commenting for possible future reference.